cpp/include/cuvs/neighbors/all_neighbors.h

```diff
@@ -0,0 +1,121 @@
+  cuvsDistanceType metric;  ///< Distance metric used for graph construction
+
+  // Algorithm-specific parameters (only one should be set based on algo)
+  cuvsIvfPqIndexParams_t ivf_pq_params;          ///< Parameters for IVF-PQ algorithm (when algo ==
+                                                 ///< CUVS_ALL_NEIGHBORS_ALGO_IVF_PQ)
+  cuvsNNDescentIndexParams_t nn_descent_params;  ///< Parameters for NN-Descent algorithm (when algo
+                                                 ///< == CUVS_ALL_NEIGHBORS_ALGO_NN_DESCENT)
+};
+
+typedef struct cuvsAllNeighborsIndexParams* cuvsAllNeighborsIndexParams_t;
+
+/**
+ * @brief Build an all-neighbors k-NN graph from a host-resident dataset (SNMG).
+ *
+ * @param[in] res             Can be a SNMG multi-GPU resources (`cuvsSNMGResources_t`)
+ * @param[in] params          Build parameters (see cuvsAllNeighborsIndexParams)
+ * @param[in] dataset         2D tensor [num_rows x dim] on host (CPU or CUDAHost)
+ * @param[out] indices        2D tensor [num_rows x k] on device (int64)
+ * @param[out] distances      Optional 2D tensor [num_rows x k] on device (float32); can be NULL
+ * @param[out] core_distances Optional 1D tensor [num_rows] on device (float32); can be NULL
+ * @param[in] alpha           Mutual-reachability scaling; used only when core_distances is provided
+ *
+ * The function partitions data into `n_clusters` clusters and assigns each row to
+ * `overlap_factor` nearest clusters. Work is balanced across GPUs according to the
+ * number of available ranks in `snmg`. Outputs always reside in device memory.
+ */
+cuvsError_t cuvsAllNeighborsBuildHost(cuvsResources_t res,
+                                      cuvsAllNeighborsIndexParams_t params,
+                                      DLManagedTensor* dataset,
+                                      DLManagedTensor* indices,
+                                      DLManagedTensor* distances,
+                                      DLManagedTensor* core_distances,
+                                      float alpha);
+
+/**
+ * @brief Build an all-neighbors k-NN graph from a device-resident dataset (single-GPU, no
+ * batching).
+ *
+ * @param[in] res             Single-GPU resources
+ * @param[in] params          Build parameters (see cuvsAllNeighborsIndexParams). For device
+ * datasets, `n_clusters` must be 1 (no batching); `overlap_factor` is ignored.
+ * @param[in] dataset         2D tensor [num_rows x dim] on device (CUDA/CUDAManaged)
+ * @param[out] indices        2D tensor [num_rows x k] on device (int64)
+ * @param[out] distances      Optional 2D tensor [num_rows x k] on device (float32); can be NULL
+ * @param[out] core_distances Optional 1D tensor [num_rows] on device (float32); can be NULL
+ * @param[in] alpha           Mutual-reachability scaling; used only when core_distances is provided
+ */
+cuvsError_t cuvsAllNeighborsBuildDevice(cuvsResources_t res,
```

Comment:
Can we please merge these into a single function? Since we're using DLManagedTensor, we can dispatch inside of the appropriate functions without having to introduce the complexity of calling a different API.